### ISSUE TITLE
[BUG] Song stops as soon as it starts on Firefox (and possibly other browsers)

### ISSUE THREAD

**Issue by longelf** [vicwomg/pikaraoke#317]
Set up PiKaraoke 1.2 as per the instructions in the readme with no issues or problems.  Running headless, can connect via browser link, it sees the sample songs I uploaded to the local folder and adds them to the queue.  All filenames match as per the requirements, nothing is zipped.  It tries to play the track, then ends with no error messages or indications of a problem.  It also does it with YouTube downloads, so I am wondering if its an ffmpeg issue.  Log is attached.

![image](https://github.com/vicwomg/pikaraoke/assets/7349359/575e5877-d814-45d0-9f91-89e797282eb0)
[Debug Log.txt](https://github.com/vicwomg/pikaraoke/files/14248499/Debug.Log.txt)

**Platform (please complete the following information):**
 - OS: Raspberry Pi OS
 - OS version: Raspberry pi Debian Bullseye 32-bit
 - Browser Firefox on Windows desktop and on Android phone, same result
 - Version [e.g. 22] 122.0.1 64-bit

**Comment by vicwomg** [comment]
Hmm interesting, looks like your ffmpeg version is old.
 `ffmpeg version 4.3.6-0+deb11u1+rpt5`

 I assume you installed using the setup.sh script?  Is there a reason you're using bullseye and not bookworm? Seems like the ffmpeg package might be out of date for that OS. If you can figure out how to install ffmpeg 6.X and above on bullseye that will probably fix this

**Comment by longelf** [comment]
I thought that might be it, but with a limited knowledge of setting up a Pi from scratch I went with the default for my Pi from the manufacturer's tool.  I did use setup.sh, and ffmpeg is apparently the latest version:

ffmpeg version 4.3.6-0+deb11u1+rpt5 Copyright (c) 2000-2023 the FFmpeg developers
built with gcc 10 (Raspbian 10.2.1-6+rpi1)

I will try Bookworm and see if that makes a difference.

**Comment by longelf** [comment]
Tried it with Bookworm, same issue...

PRETTY_NAME="Raspbian GNU/Linux 12 (bookworm)"
NAME="Raspbian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=raspbian
ID_LIKE=debian

pi@pikaraoke:~ $ sudo apt install ffmpeg -y
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ffmpeg is already the newest version (8:5.1.4-0+rpt1+deb12u1).
0 upgraded, 0 newly installed, 0 to remove and 94 not upgraded.

Not v6 of ffmpeg but I don't know how to upgrade it as the apt command seems to think its the latest.

pi@pikaraoke:~ $ ffmpeg -version
ffmpeg version 5.1.4-0+rpt1+deb12u1 Copyright (c) 2000-2023 the FFmpeg developers
built with gcc 12 (Raspbian 12.2.0-14+rpi1)
configuration: --prefix=/usr --extra-version=0+rpt1+deb12u1 --toolchain=hardened --incdir=/usr/include/arm-linux-gnueabihf --enable-gpl --disable-stripping --disable-mmal --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libglslang --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librist --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sand --enable-sdl2 --disable-sndio --enable-libjxl --enable-neon --enable-v4l2-request --enable-libudev --enable-epoxy --enable-pocketsphinx --enable-librsvg --enable-libdc1394 --enable-libdrm --enable-vout-drm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-libplacebo --enable-librav1e --enable-shared --libdir=/usr/lib/arm-linux-gnueabihf --cpu=arm1176jzf-s --arch=arm
libavutil      57. 28.100 / 57. 28.100
libavcodec     59. 37.100 / 59. 37.100
libavformat    59. 27.100 / 59. 27.100
libavdevice    59.  7.100 / 59.  7.100
libavfilter     8. 44.100 /  8. 44.100
libswscale      6.  7.100 /  6.  7.100
libswresample   4.  7.100 /  4.  7.100
libpostproc    56.  6.100 / 56.  6.100

**Comment by vicwomg** [comment]
I think I'm wrong about ffmpeg, just fired up my pi and it's running this without issue: 
```
pi@pikaraoke:~ $ ffmpeg --version
ffmpeg version 4.3.6-0+deb11u1+rpt5 Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 10 (Raspbian 10.2.1-6+rpi1)
```
Really, the existence of `librubberband` is what we want. Looks enabled on yours.

Looks like this is actually a firefox issue. I was able to repro on firefox, but not chrome or safari. Hard to tell what's causing it off the top of my head. I will update the issue description. In the meantime, can you try another browser?

**Comment by longelf** [comment]
Tested with Chrome and Edge, same issue, so if you're getting a problem with Firefox I would have to assume its a separate issue.  I'll play around some more and see if I can figure it out.

**Comment by vicwomg** [comment]
Very weird. I'll see if I can find time to fire up my PC to test

**Comment by longelf** [comment]
Small update, it does it in headered mode connected directly to the TV via HDMI, so not an audio output/headless mode problem either :(

Feel free to hit me up with any troubleshooting or testing requirements, I really want to get this working, it's the closest I have found to the right solution for my needs.

**Comment by longelf** [comment]
Starting to think it's a problem with the Pi 3 itself, I have a Pi 5 on order, will test with that

**Comment by vicwomg** [comment]
It’s possible, I run on a pi 3b myself though without issue.

You might also want to try installing to your windows machine to see if
that works. It’s possible this is a LAN thing

Sorry, haven’t had much time to investigate

**Comment by longelf** [comment]
Brand new Pi 5, brand new 128GB Samsung SD Card, OS pre-installed and latest PiKaraoke installed as per instructions in readme.md:

sudo apt-get update
sudo apt-get install git
git clone https://github.com/vicwomg/pikaraoke.git
cd pikaraoke
./setup.sh
mkdir /home/pi/.config/autostart
nano /home/pi/.config/autostart/pikaraoke.desktop

[Desktop Entry]
Type=Application
Name=Pikaraoke
Exec=/home/pi/pikaraoke/pikaraoke.sh

Copied over confirmed working Karaoke CDG/MP3 pairs, not zipped, connected via HDMI to TV running headered.

Tried adding songs using Chromium directly on the Pi, Firefox and Microsoft Edge, exact same issue.  Songs start then stop immediately.  I am at a loss.  Not sure how it could be a network issue as the Karaoke files themselves are local to the Pi.  

Here are the pertinent software versions:

Microsoft Edge Version 121.0.2277.128 (Official build) (64-bit), no plugins installed
Mozilla Firefox 122.0.1 (64-bit)

Linux raspberrypi 6.1.0-rpi8-rpi-2712 #1 SMP PREEMPT Debian 1:6.1.73-1+rpt1 (2024-01-25) aarch64

Python 3.11.2

ffmpeg version 5.1.4-0+rpt2+deb12u1 Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 12 (Debian 12.2.0-14)
  configuration: --prefix=/usr --extra-version=0+rpt2+deb12u1 --toolchain=hardened --incdir=/usr/include/aarch64-linux-gnu --enable-gpl --disable-stripping --disable-mmal --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libglslang --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librist --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sand --enable-sdl2 --disable-sndio --enable-libjxl --enable-neon --enable-v4l2-request --enable-libudev --enable-epoxy --libdir=/usr/lib/aarch64-linux-gnu --arch=arm64 --enable-pocketsphinx --enable-librsvg --enable-libdc1394 --enable-libdrm --enable-vout-drm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-libplacebo --enable-librav1e --enable-shared
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100

Will try a Windows install at some point, but for now I am setting this to one side.

**Comment by longelf** [comment]
More information in case it helps, I now have YouTube downloads working just fine, still can't get it to play a CD+G.  I have been autostarting it on the Pi as per the install instructions, but wanted to try manually starting it in case that was anything to do with it, and this is the output I get:

`pi@pikaraoke:~/pikaraoke $ ./pikaraoke.sh
[2024-03-16 14:23:16] INFO: Fetching available songs in: /home/pi/pikaraoke-songs/
[16/Mar/2024:14:23:17] ENGINE Bus STARTING
[2024-03-16 14:23:17] INFO: [16/Mar/2024:14:23:17] ENGINE Bus STARTING
[16/Mar/2024:14:23:17] ENGINE Serving on http://0.0.0.0:5555
[2024-03-16 14:23:17] INFO: [16/Mar/2024:14:23:17] ENGINE Serving on http://0.0.0.0:5555
[16/Mar/2024:14:23:17] ENGINE Bus STARTED
[2024-03-16 14:23:17] INFO: [16/Mar/2024:14:23:17] ENGINE Bus STARTED
Traceback (most recent call last):
  File "/home/pi/pikaraoke/app.py", line 866, in <module>
    driver = webdriver.Chrome(service=service, options=options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/chromium/webdriver.py", line 61, in __init__
    super().__init__(command_executor=executor, options=options)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 208, in __init__
    self.start_session(capabilities)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 292, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 347, in execute
    self.error_handler.check_response(response)
  File "/home/pi/pikaraoke/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.SessionNotCreatedException: Message: session not created: Chrome failed to start: exited normally.
  (session not created: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/chromium-browser is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
Stacktrace:
#0 0x55561f6ec3e4 <unknown>
#1 0x55561f421ac8 <unknown>
#2 0x55561f450fac <unknown>
#3 0x55561f44dc44 <unknown>
#4 0x55561f48cf08 <unknown>
#5 0x55561f458d74 <unknown>
#6 0x55561f459760 <unknown>
#7 0x55561f6b6a74 <unknown>
#8 0x55561f6b9e04 <unknown>
#9 0x55561f6a6d7c <unknown>
#10 0x55561f6ba714 <unknown>
#11 0x55561f68f0e0 <unknown>
#12 0x55561f6dbbc4 <unknown>
#13 0x55561f6dbdc0 <unknown>
#14 0x55561f6eb5b4 <unknown>
#15 0x7ffef499ee58 <unknown>
#16 0x7ffef4a07f9c <unknown>`

Chromium fails to launch.

**Comment by longelf** [comment]
Checking the tmp folder, I can see the cdg/mp3 pair so it is unzipping fine, seems it fails after that.  Could it be a permissions issue?  I tried changing the tmp/pikaraoke to 777 with recursion to get any sub-folders, but it didn't help, and I can't find a log file anywhere to give me any further clues.  I can also only get it to auto-start after a reboot, it fails using ./pikaraoke.sh as indicated above, and I don't understand why there's no log file anywhere either.  This is another completely clean install with all the latest versions of everything, following the install instructions, I just can't figure out why it isn't working.  

@vicwomg You say yours is fine, is that after updating to 1.2 from a previous version, or with a brand new, clean setup?

**Comment by longelf** [comment]
FYI @vicwomg I have it working perfectly using my Pi3 with v1.1.2, using the image on the releases page.  Can't test with the Pi5 as the 1.1.2 image isn't compatible with it.  I also had to chmod the default download folder to 777 so I could copy files to it.  So, to recap, clean install of 1.1.2 on a Pi3 works perfectly, clean install of 1.2 on a Pi3 or Pi5 always results in the same issue, CDG/MP3 files fail to play, stopping as soon as they start.

**Comment by vicwomg** [comment]
> FYI @vicwomg I have it working perfectly using my Pi3 with v1.1.2, using the image on the releases page. Can't test with the Pi5 as the 1.1.2 image isn't compatible with it. I also had to chmod the default download folder to 777 so I could copy files to it. So, to recap, clean install of 1.1.2 on a Pi3 works perfectly, clean install of 1.2 on a Pi3 or Pi5 always results in the same issue, CDG/MP3 files fail to play, stopping as soon as they start.

1.1.2 uses a completely different streaming mechanism. Basically just uses vlc, so any success there really wont help here.

Last week I created a fresh install from scratch on a pi3. No problems with the standard install from scratch.

I think I need to hear more about your setup. Is your pi connected to the TV? Or are you hosting on the pi, and opening the splash screen on another computer? Browser? All relevant info.

A few things looking at your last few comments:
- The autostart instructions assume you are running with your pi connected to the TV and hosting the splash screen on the pi. If you're not, you should run with --headless to remove that chromium error
- What browser are you using for your splash screen? I'm finding that not all browsers like transcoded ffmpeg streams (cdg files will be transcoded. So will non-mp4 video formats). Have you tried chrome?

**Comment by longelf** [comment]
@vicwomg Here is what I am doing....

Brand new Pi5 8gb with a 128gb Samsung SD Card (same result with the Pi3, which is why I bought the Pi5 for testing)
Using the Raspberry Pi Imager software, download and flash the latest OS - 2023-12-05-raspios-bookworm-arm64-full.img (I have tried many different versions, 32 and 64 bit, all yield the same result).
When flashing, I have the Imager program applying custom settings to set up the WiFi, enable SSH and set the default user as "pi" with a password.
Connect the Pi to the TV via HDMI0 and let it run through its boot up stuff.
Once up and running, with the OS Desktop showing on the TV, SSH in from a Windows 10 machine and run the following commands:

sudo apt-get update
sudo apt-get install git
git clone https://github.com/vicwomg/pikaraoke.git
cd pikaraoke
./setup.sh

No errors are reported through this process.
I then try to run ./pikaraoke.sh but get the errors I listed above and the splash screen does not appear as Chromium fails to launch.
Then I set it up to autostart with the following:

mkdir /home/pi/.config/autostart
nano /home/pi/.config/autostart/pikaraoke.desktop

[Desktop Entry]
Type=Application
Name=Pikaraoke
Exec=/home/pi/pikaraoke/pikaraoke.sh

Not set to run headless as it is connected to the TV, which is where I want the video output to be, adding songs from other devices.
Write the file out and reboot, Chromium loads and the PiKaraoke splash screen loads on the TV, all good so far.

Connect to http://192.168.1.145:5555/ on any browser (the IP address of the Pi), and I have tried many different browsers on different devices, and I can set up my username all ready to go.
I connect to the Pi via WinSCP and copy across some CDG/MP3 zip files that I know work, and work on the Pi3 in 1.1.2.
Rescan the song directory and the files appear.  
Add one to the queue, the playback controls show briefly in the Home tab, then disappear immediately.
If I search for a song on YouTube, it downloads and plays without an issue.

As I mentioned before, I can't find a log file, Chromium fails to load from the command line, but I do see the unzipped files in the temp folder.

There is nothing else I am doing besides these steps.

**Comment by vicwomg** [comment]
Your first-time chromium fails to launch because you're doing it over SSH and it's trying to open it on the client you're sshing from which wont work. To launch on the target pi, you need to do this instead:

`DISPLAY=:0.0 ./pikaraoke.sh`

If you want logging, modify your autolaunch command to or similar

`Exec=/home/pi/pikaraoke/pikaraoke.sh -l10  > /tmp/pikaraoke.log`

I'm afraid to say, just flashed a fresh 3/12/24 Bullseye 32 to raspberry pi and cannot reproduce the cdg issue as you described. They play fine. The CDGs I tried are from here: https://archive.org/download/KaraokeVocalStarsPartyHits

To me, this all sounds like a ffmpeg transcoding issue. mp4s from youtube are supported natively by the browser are not transcoded in real time as they are streamed to the browser, but cdgs are. So ffmpeg is streaming the bytes of the transcoded file in realtime to the browser as fast as it will take them.

I will say that on pi3, the ffmpeg version is: `ffmpeg version 4.3.6-0+deb11u1+rpt5`. I cannot speak for the compatibility default version on pi5 as I don't have one. You may want to experiment with forcing it to download the same version as mine.

Sorry to hear you've been burning so much time on this, but I simply can't reproduce the issue so I'm at a loss. Maybe more folks can chime in if they are experiencing the same.

**Comment by longelf** [comment]
Thanks Vic for the help trying to navigate through this and I agree, something in my setup is where the issue is but like yourself I'm at a loss to explain it.  I will continue to play with it, your latest insights will be invaluable in that process, and if I get it working I'll let you know what I did different.  This setup is perfect for my use case, and not finding a way out of this issue is so annoying as a result, so keep up the amazing work, one day I'll get it up and running.

**Comment by longelf** [comment]
@vicwomg Fixed it, now it works like a charm...

Clean flash of 3/12/24 Bullseye 32 (which I had tried before), but this time no custom settings in the Pi Imager.  Booted the Pi3, and manually set up WiFi and the pi user with a connected keyboard, not using ssh at this point.  Used the Pi's terminal to install git and then your repository, installed PiKaraoke 1.2 and setup autostart.  Rebooted and confirmed it was downloading Youtube songs, worked great.  Enabled ssh and transferred some songs over via WinSCP, PiKaraoke plays them with no issue.

The only things I did differently were not using the custom network settings in the Pi Imager, setting them up manually with a connected keyboard, and not using ssh to issue commands to the Pi, only to transfer CDG/MP3 files via WinSCP.

In conclusion... Yay!  I am now super happy.

**Comment by longelf** [comment]
Quick update, did exactly the same as above but with the Pi5, which isn't compatible with Bullseye so I had to use Bookworm, and it doesn't work, so I'm thinking Bullseye is required until PiKaraoke can be made compatible with Bookworm.

**Comment by vicwomg** [comment]
I developed most of this on pi4 bookworm, so that's still strange. I'll have to try with the latest on a pi4

**Comment by vicwomg** [comment]
I **can** reproduce this in the latest pi4 something has changed in bookworm since the initial 1.2 release.

I will close out this ticket and open a more focused task
